### ISSUE TITLE
Support dependent pull requests

### ISSUE THREAD

**Issue by Artur-** [Reviewable/Reviewable#253]
Some times, e.g. when refactoring a lot, it's very useful to make many pull requests which are dependent on each other instead of one large pull request. It's much easier for the reviewer if PR1 is rename "A to B and C to D" and PR2 is "Fix bad code in B and D". If you do this with Reviewable today, you will end up with the "rename" changes also in the review of PR2, which is not what you want. 

It would be nice if you could create branch1 and branch2 for PR1 and PR2 where branch2 is based on branch1, and still only see PR2 code in the review of PR2.

**Comment by pkaminski** [comment]
Hey Arthur, this is already supported albeit requiring a bit more work than
would be ideal. When you create the PR2 you need to specify branch1 as the
target. You'll then get a clean diff for your review. Once both PRs are
reviewed, you'll want to merge PR2 first, or manually merge branch2 into
master, since GitHub doesn't allow changing a PR's target. Unfortunately
GitHub doesn't make dependent / sequential feature development easy.

**Comment by pkaminski** [comment]
@Artur- did the reply above help or are you still looking for something more?

**Comment by Artur-** [comment]
The workaround with different branches probably works for simple cases. 

If PR2 has been merged into branch1 and you then notice that you have conflicts in branch1 and need to resolve those before merging into master, you will likely not have a good time. Then you would get a new review with PR1+PR2+merge resolving changes.

If you add a branch 3/PR3, I suspect you will have quite a mess.

Would it be possible to see that PR2 is based on PR1 and just remove PR1 files from the review of PR2? Basically the same thing which happens after PR1 has been merged and the changes are in master.

If you have PR3 -> PR2 -> PR1, then each would only show the difference to the previous pull request instead of the difference to the target branch.

**Comment by pkaminski** [comment]
I took some time to dig into the code and I think this would be feasible.  I'd also want to add a status check to the dependent PR to prevent (or at least warn against) merging it in before the "base" PR is merged itself.  I'll add it to the todo list.

**Comment by pcorpet** [comment]
:+1: this would be awesome @pkaminski

**Comment by erikwright** [comment]
I also consider this a potential key differentiator from GitHub. For those of us who drink the Kool Aid of many small commits, not being able to pipeline is a humungous bottleneck.

**Comment by pkaminski** [comment]
Does the new GitHub feature to [change the base branch of a PR](https://github.com/blog/2224-change-the-base-branch-of-a-pull-request) obviate this feature request?

If so, then a secondary issue is how Reviewable should react to a base branch change.  Currently, changes in the base branch don't trigger the creation of new revisions / diffs, to avoid making a mess of the review as other PRs get merged into the base branch.  The assumption is that the diff against the PR's base at the time a revision's commits were pushed is sufficient as long as the PR remains mergeable, and if it becomes unmergeable then the author will need to push new commits to head anyway (hence creating new revision against the current base).  Should this also apply to base branch changes, or should such a change trigger the creation of new revisions?  (Note that this could potentially be a large number if the review style is "review each commit".)

A tertiary issue is that currently GitHub doesn't fire an event when the base branch is changed, but I'm hoping that's just a temporary omission.

**Comment by pkaminski** [comment]
Given the lack of feedback I've decided to treat a base branch change like any other base commit change, by doing nothing.  I also integrated the base branch change functionality into Reviewable (works exactly the same as in GitHub, but it's one click closer), which should let you chain PRs pretty easily.  Closing.

**Comment by NicoK** [comment]
Unfortunately, the different base branches workaround only works for PRs inside the same repository.

Suppose, I forked `apache/flink` into my own repository `NicoK/flink` and then develop there. By trying to make PRs atomic, small and simple, I create PR1 for part `A` and PR2 for parts `A+B` both into `apache/flink` (since I want a review from them, not in my repository).
By this, I stack PR2 onto PR1 and the reviewer may first review (& merge) PR1 and then go on with PR2. It is, however, only until the merge of PR1 that he can review the changes of `B` directly, i.e. excluding the ones from `A` (depending on whether a rebase is required).

**Comment by pkaminski** [comment]
You're right, this method is limited to same-repo branches and I can't think of a workaround.  :(  However, given that implementing this directly in Reviewable would be tricky and GitHub's current solution addresses a large chunk of the problem space, it's unlikely I'll add this feature soon.  Nonetheless, I'll reopen this issue to at least give it some more visibility and see if it's still impacting many people.

**Comment by jwnimmer-tri** [comment]
For the teams I work with, there is definitely pent-up demand for having multiple PRs open at once that linearly build on each other, with the goal that code review covers the unique additions in that PR above and beyond its ancestors, and that merging later PRs would be contingent on their ancestors merging first.  (I _think_ that is what you're asking about in this issue, but please correct me if I'm wrong.)

For now, we are working around it by slowing down our PR trains to only have one or two PRs open at a time from each train (an manually explaining which part of the second PR to ignore), or occasionally using "review each commit separately" to kind of box them all up into a single review (with careful attention to the file matrix to walk the mated revisions).

I have not yet tried working around it by opening personal-repository PRs (and changing base branches as they float to the top), but perhaps that is another viable work-around?  I would expect that to really lose out on the "watching" firehose for repositories though, since discussions would post into my personal fork, until the PR is at the head of the train?

Aside and apart from the base-branch model that github supports, if reviewable had a "based-on" metadata that cited another PR as its base _for the purposes of code review only_, and where it had a mandatory interlock that the full diff must be showing prior to merging (so that the final form of the code was reviewed, achieved when the other PR(s) merge), that would be extremely valuable for us, and I think a stand-out feature of the tool vs its competition.

**Comment by cody-somerville** [comment]
It would be great if this worked similar to Gerrit.

**Comment by pkaminski** [comment]
This is steadily floating up the priority list:  I hear the demand, and it's a good fit for the new product direction of "serious code reviews".  That said, 1) there are still much higher priority things ahead of it that impact the core review experience, and 2) I'm concerned that GitHub will go ahead and implement this feature a couple months after I get around to it (like they did with multi-assignees) and cause me major migration headaches again.  So not right now... but a definite maybe in the future.

**Comment by fwouts** [comment]
@pkaminski Any updates since March?

We're looking to switch to a better code review tool. We're currently using GitHub directly, which makes it really hard to split up large code changes into small, atomic pull requests. Being able to diff against a particular branch, without changing its base branch, is a required feature for the tool we'll pick.

**Comment by pkaminski** [comment]
Hey François, no real updates.  This feature is still important, but the current rewrite into Vue (and ditching of AngularJS) takes precedence, and will take a while longer.

**Comment by fwouts** [comment]
OK, understand. Do you have an ETA for the rewrite to be done?

**Comment by pkaminski** [comment]
It's still early days so the velocity is pretty inconsistent as we randomly run into "interesting" issues to resolve.  I don't expect it to be done before March, though, and likely no later than June.  Just guesses, though, sorry.

I take it the same-repo "change the target branch" workaround isn't acceptable to you?

**Comment by fwouts** [comment]
It's not ideal as it requires manual work from the developer (and remembering to switch it back before merging).

But maybe that can be alleviated by building a small CLI tool leveraging GitHub's API to help manage multiple PRs where one depends on another. GitHub's model of 1 branch = 1 PR is what makes these things hard, I don't blame Reviewable for that 😊

**Comment by hazelnusse** [comment]
Hi @pkaminski any update here?

**Comment by hazelnusse** [comment]
Useful reference material on how this works in Gerrit can be found here:
https://docs.google.com/presentation/d/1C73UgQdzZDw0gzpaEqIC6SPujZJhqamyqO1XOHjH-uk/edit#slide=id.g4d6c16487b_1_1711

Basically, that ability for a developer to work on a single local branch, with 1 or more commits, each of which result in an independent review when pushed. The key idea is to avoid the need for developers to manage separate branches for each of the intermediate commits.

I suspect that to implement this, something akin to Gerrit's Change-Id will be needed, even if it doesn't appear directly in the commit message (although making it explicit in the commit message does have it's benefits). Details of how the commit message hook generates the Change-Id can be found here:
https://gerrit-review.googlesource.com/Documentation/cmd-hook-commit-msg.html#_implementation

**Comment by pkaminski** [comment]
So as a general update:  the Vue migration is ongoing, and of course I underestimated how long it would take.  The good news is that we've now reached the hardest part -- the diff view itself.  I'm going to hold off on predicting when that will be done, given my track record so far.  :)

I would like to discuss @hazelnusse's last comment above, though.  What you seem to be proposing there is simply a commit-by-commit review:  one branch, a stack of commits, and rebasing to post new changes.  Reviewable already supports this if you switch the review mode to the "review each commit separately" review style, and will do its best to match up commits between rebases by their summary.  (This could admittedly be improved by matching on Change-Id lines when available as suggested in https://github.com/Reviewable/Reviewable/issues/284, but I don't think it's a big pain point especially if you amend-commit without changing the summary.)  The only limitation is that you must merge the whole branch instead of individual commits unless you resort to manual cherry-picking.

What's proposed here is having a sequence of branches, each with multiple commits (that may not be reviewed individually, i.e., "combine commits for review" style), such that each branch in the stack can be merged in as it's completed.  Does that tally with your understanding or are we talking about different things?

**Comment by hazelnusse** [comment]
What I'm proposing is that a single branch with N commits yield N individual PRs (not one big PR with the option to review each commit separately). Each of the N PRs gets updated when a rebase and subsequent push of the branch occurs, including any edits if they occurred during the rebase (`rebase -i HEAD~5`, edit a couple of commits, then push). If no edits occurred to the commit, but the SHA changed because the parent SHA changed, a new revision with zero file diffs would occur and there would be no need for reviewers of the intermediate commit to re-examine the intermediate commit PR.

The key idea is that the developer works on 1 branch, with multiple commits, and each commit maps to its own PR.

**Comment by pkaminski** [comment]
OK, got it.  This would be non-trivial to implement since GitHub requires a distinct branch per PR, so Reviewable would need to generate N branches, open up N PRs, and keep everything synced up as you push to the main branch.  TBH it's also a pretty bad fit for Reviewable's data structures -- normally everything revolves around the review / single PR, whereas for this I'd need something one level "up".  So yeah, not impossible by any means, but a significant effort.

Why are you averse to the one PR with multiple commits approach?  Reviewable already matches up commit pairs for diffing, so if you rebase and only some commits changed you'll only see diffs for those.  The only downside I'm aware of is that you can't merge things in one commit at a time -- is that critical for your needs?

**Comment by hazelnusse** [comment]
If multiple commits are part of a single PR, as I understand it, CI only builds the last commit, so there is no guarantee that intermediate commits don't cause breakages if they are merged to the target branch. This can cause pain and suffering with git bisect. This can be fixed if there is a policy of squashing prior to merging, but then either the developer has to to do this (and potentially introduce something that hasn't been reviewed, or reviewers must be required to review post squash) or the tool has to do it and figure out how to nicely combine commit messages (not generally possible) -- either way, this isn't ideal if the commits should be kept separate instead of squashed.

**Comment by pkaminski** [comment]
Thanks, good points.  For CI, a quick search show that both Travis and CircleCI are or can be configured to build every commit.  This would make sense, since in the GitHub API you report status checks per commit SHA, rather than per branch.  Which CI are you using?

But if you do end up needing to squash, you can constrain GitHub's merge flavors to require squashing, and customize your Reviewable review completion condition to output a custom `defaultSquashCommitMessage`.  (By default, Reviewable will use the PR's summary line, and list all the commit summaries as a bulleted list in the body, which may be sufficient.)

**Comment by hazelnusse** [comment]
We are using Jenkins CI.

**Comment by pkaminski** [comment]
I'm not personally familiar with Jenkins, and based on some quick searching it looks like build-per-commit might be somewhat challenging to achieve on this platform.  FWIW, I found these [very straightforward instructions](https://oncomputingwell.princeton.edu/2018/01/triggering-a-jenkins-build-every-time-changes-are-pushed-to-a-git-branch-on-github/), but I suspect they're too simple to actually do the job.  Beyond that, there are hints around triggering builds parametrized by the commit SHA, and a [12 year old feature request](https://issues.jenkins-ci.org/browse/JENKINS-673) that's still open...

In any case, I don't think it's unreasonable to expect a modern CI system to be able to build every commit.

**Comment by pkaminski** [comment]
GitHub just launched [pull request retargeting](https://github.blog/changelog/2020-05-19-pull-request-retargeting/):  when a branch gets merged, any pull requests that were targeting it will now target the merged branch's base instead.  I think this should make stacked/dependent PRs work pretty smoothly (unless you're using forked repos).  At this point, I believe this hits the point of "close enough" where it's not worth competing with GitHub's functionality.  Thoughts?

**Comment by pcorpet** [comment]
Yep, it sounds good enough for me. Thanks for pointing that out.
